File: component/db/Class_dbkeyclick.pas
unit Class_dbkeyclick;

interface

uses
  mysql.data.mysqlclient,
  system.configuration,
  system.data;

type
  TClass_dbkeyclick = class abstract
  private
    { Private Declarations }
  protected
    connection: mysqlconnection;
    procedure Close;
    procedure Open;
  public
    constructor Create;
  end;

implementation

constructor TClass_dbkeyclick.Create;
begin
  inherited Create;
  // TODO: Add any constructor code here
  connection := mysqlconnection.Create;
  connection.ConnectionString := configurationmanager.AppSettings['dbkeyclick_connection_string'];
end;

procedure TClass_dbkeyclick.Close;
begin
  connection.Close;
end;

procedure TClass_dbkeyclick.Open;
begin
  if connection.State <> connectionstate.OPEN then begin
    connection.Open;
  end;
end;

end.

File: component/db/Class_dbkeyclick.cs
using MySql.Data.MySqlClient;
using System.Configuration;
using System.Data;

namespace Class_dbkeyclick
  {

  public abstract class TClass_dbkeyclick
    {

    protected MySqlConnection connection = null;

    public TClass_dbkeyclick() : base()
      {
      connection = new MySqlConnection(connectionString:ConfigurationManager.AppSettings["dbkeyclick_connection_string"]);
      }

    protected void Close()
      {
      connection.Close();
      }

    protected static void ExecuteOneOffProcedureScriptWithTolerance
      (
      string procedure_name,
      MySqlScript my_sql_script
      )
      {
      var done = false;
      while (!done)
        {
        try
          {
          my_sql_script.Execute();
          done = true;
          }
        catch (MySqlException the_exception)
          {
          if (the_exception.Message != "PROCEDURE " + procedure_name + " already exists")
            {
            throw;
            }
          }
        }
      }

    protected void Open()
      {
      if (connection.State != ConnectionState.Open)
        {
        connection.Open();
        }
      }

    }

  }
